### PR TITLE
HTTP proxy support + some cleanup

### LINKED ISSUE #8114: ACLK, challenge/response and Claiming should support HTTP proxies
##### Feature idea summary
MQTT over WSS should work over HTTP proxy

- [x] read `http_proxy` env var and config value if `proxy` set to `env`
- [x] use proxy from config if set to `http://HOST:PORT`
- [x] test with squid proxy

### PR DESCRIPTION
**Pull request by underhood**
##### Summary
Fixes #8114

##### Component Name
aclk, claiming, challenge/response
##### Test Plan
With following in the config file:
```ini
[cloud]
        cloud base url = https://192.168.56.1:8443

[agent_cloud_link]
        proxy = http://192.168.56.101:3128
```

In my case `192.168.56.101:3128` is squid proxy running in VM. `192.168.56.1` is machine connectible trough aforementioned proxy where local cloud environment is running. It is recommended using firewall to block access to cloud environment without proxy (or check with wireshark). Additional way to check connection is indeed made trough proxy is checking squids `access.log` where you should see something like `192.168.56.1 TCP_TUNNEL/200 2188 CONNECT 192.168.56.1:8443 - HIER_DIRECT/192.168.56.1 -`.

Squid proxy has to be correctly configured (if something doesn't work check `access.log` to check it was denied by PROXY). Example of important keys:
```
acl localnet src 192.168.0.0/16 # To allow connections to proxy from agent IP (connections to proxy)
acl SSL_ports port 8443 # To allow TLS upgrade if target port (from proxy) is 8443 - by default only 443 allowed to use CONNECT to upgrade to TLS
acl SSL_ports port 9009 # Because currently claiming doesn't work over 8443 (even without proxy) for some reason
acl Safe_ports port 8443 # To allow connections to target port (from proxy) 8443
acl Safe_ports port 9009 # Because currently claiming doesn't work over 8443 (even without proxy) for some reason
http_access allow localnet
http_access allow localhost
# Squid normally listens to port 3128
http_port 3128
```
Please consult `access.log` of squid proxy for troubleshooting as your distribution defaults for squid config might be different. In case needed I can send you copy of my config. For example settings above rely on some defaults on ubuntu install of squid like:
```
# Deny requests to certain unsafe ports
http_access deny !Safe_ports

# Deny CONNECT to other than secure SSL ports
http_access deny CONNECT !SSL_ports
```

Tests:
- Claiming works trough HTTP proxy (currently with 9009)
- Challenge & response works trough HTTP proxy
- Test ACLK link gets established.

Example of log where it is working correctly (from my testing):
```
2020-03-17 20:22:53: netdata INFO  : ACLK_Main : Waiting for netdata to be claimed
2020-03-17 20:22:53: netdata INFO  : ACLK_Main : Attempting to establish the agent cloud link
2020-03-17 20:22:53: netdata INFO  : ACLK_Main : Retrieving challenge from cloud: 192.168.56.1 8443 /api/v1/auth/node/946da25e-67a9-11ea-a233-3085a98d4bee/challenge
2020-03-17 20:22:53: netdata INFO  : ACLK_Main : aclk_send_https_request GET
2020-03-17 20:22:53: netdata INFO  : ACLK_Main : Disabling SSL certificate checks
2020-03-17 20:22:53: netdata INFO  : ACLK_Main : Connecting using HTTP proxy:"http://192.168.56.101:3128"
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : aclk_send_https_request POST
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Disabling SSL certificate checks
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Connecting using HTTP proxy:"http://192.168.56.101:3128"
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Processing callback LWS_CALLBACK_PROTOCOL_INIT
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Connecting using HTTP proxy:"http://192.168.56.101:3128"
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Disabling SSL certificate checks
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Processing callback LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Using challenge-response: 946da25e-67a9-11ea-a233-3085a98d4bee / +uTbXJ2XFrmI0Kerm3Etk5WR0p8=
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Processing callback LWS_CALLBACK_CLIENT_ESTABLISHED
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Establishing MQTT link to [192.168.56.1:8443]
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Connection to cloud estabilished
2020-03-17 20:22:54: netdata INFO  : ACLK_Main : Connection detected (0 queued queries)
```

### CODE REVIEW

**Comment by amoss** [comment]
This doesn't match the coding style. Please run it through `clang-format` using the file in the root of the repo.

File: aclk/aclk_common.c [hunk 1]
```diff
@@ -24,124 +39,162 @@ static inline ACLK_PROXY_TYPE aclk_find_proxy(const char *string)
...
+    static ACLK_PROXY_TYPE proxy_type = PROXY_NOT_SET;
+
+    if (proxy_type == PROXY_NOT_SET)
+        proxy = aclk_lws_wss_get_proxy_setting(&proxy_type);
+
+    *type = proxy_type;
+    return proxy;
+}
+
 int aclk_decode_base_url(char *url, char **aclk_hostname, char **aclk_port)
```
Comment (stelfrag) [inline aclk/aclk_common.c]: In this function I put the following config by accident (trailing /) 
```
cloud base url = https://dev.private.netdata.cloud/
```

and it fails to parse correctly 

```
2020-03-20 10:27:27: netdata INFO  : ACLK_Main : (0200@aclk/aclk_:aclk_decode_bas): Setting ACLK target host=(null) port=(null) from https://dev.private.netdata.cloud/
```

Comment (underhood) [reply aclk/aclk_common.c]: Thanks guess decode_base_url from challenge response has bug. I will check

**Comment by underhood** [comment]
I rebased on latest master

File: aclk/aclk_lws_wss_client.c [hunk 2]
```diff
@@ -208,42 +208,81 @@ void aclk_lws_wss_client_destroy()
 #endif
 }
 
-int aclk_wss_set_socks(struct lws_vhost *vhost, const char *socks) {
+static int aclk_wss_set_socks(struct lws_vhost *vhost, const char *socks)
+{
     char *proxy = strstr(socks, ACLK_PROXY_PROTO_ADDR_SEPARATOR);
 
-    if(!proxy)
+    if (!proxy)
```
Comment (amoss) [inline aclk/aclk_lws_wss_client.c]: There is unnecessary duplication in the code. This code path cannot be reached unless aclk_find_proxy found the appropriate prefix, then code is being run to check the prefix again.

File: aclk/aclk_common.c (hunk 1 above)
Comment (amoss) [reply aclk/aclk_common.c]: That won't work on https://host:1234/ - there needs to be an `else` to make the new case distinct.

**Comment by underhood** [comment]
I think I missunderstand something, seems to parse `https://host:1234/` well here

**Pull request merged**